NEXUS-4257, Catch unknown session exception when logging in the anon user

bdemers opened the pull request.

bdemers committed:
NEXUS-44257, NEXUS-4257
Catch unknown session exception when logging in the anon user, which seems to be caused by the DelegatingSession
If an unknown session is thrown, the subject will be logged out, then a second login attempt will be made.

nabcos commented:
commit is missing useful log message

nabcos commented on nexus/nexus-rest-api/src/main/java/org/sonatype/nexus/security/filter/authc/NexusHttpAuthenticationFilter.java:
```diff
@@ -235,15 +236,31 @@ public class NexusHttpAuthenticationFilter
+            Session anonSession = subject.getSession( false );
 
-            if ( getLogger().isDebugEnabled() )
+            this.getLogger().warn( "UnknownSessionException was thrown for with session: '{}'", anonSession, e );
```
- typo ("for with")
- log message could be more clear ("Unknown session while logging in anonymous" or sth) because warn is visible with the default log settings
- should this be warn level anyway?

bdemers replied:
You might be right about the log level.  I personally would like to see it, but I don't think the customer wants to see a warning, for something that is recoverable.

peterlynch commented on nexus/nexus-rest-api/src/main/java/org/sonatype/nexus/security/filter/authc/NexusHttpAuthenticationFilter.java:
```diff
@@ -219,11 +221,10 @@ public class NexusHttpAuthenticationFilter
-        {
-            getLogger().debug( "Attempting to authenticate Subject as Anonymous request..." );
-        }
+        getLogger().debug( "Attempting to authenticate Subject as Anonymous request..." );
```
couldn't this log message be more informative if the subject information was included in the message?

bdemers replied:
In this case the subject will always be anonymous, and subject.getPrincipal() would be null. (until after the login)

peterlynch commented on nexus/nexus-rest-api/src/main/java/org/sonatype/nexus/security/filter/authc/NexusHttpAuthenticationFilter.java:
```diff
@@ -307,7 +330,7 @@ public class NexusHttpAuthenticationFilter
                 getSubject( request, response ).logout();
             }
-            catch ( SessionException e )
+            catch ( SessionException e ) //TODO: investigate why this is getting thrown
```
adding reference the issue number on this line would prove useful for the future

velo commented on nexus/nexus-rest-api/src/main/java/org/sonatype/nexus/security/filter/authc/NexusHttpAuthenticationFilter.java:
```diff
@@ -219,11 +221,10 @@ public class NexusHttpAuthenticationFilter
 
     protected boolean executeAnonymousLogin( ServletRequest request, ServletResponse response )
     {
-        if ( getLogger().isDebugEnabled() )
```
Why removing the getLogger().isDebugEnabled() check?  Just curious

bdemers replied:
There is not string concatenation.  isDebugEnabled() is better used when doing something like:
log.debug("something: "+ myObject);
or
log.debug("something: "+ someMethodCall() );

bdemers commented:
@nabcos, good call, I'll fix the commit message when i merge

bdemers committed:
NEXUS-4267 NEXUS-4257 cleaning up log message, and adding jira issue reference to TODO

nabcos commented:
+1

bdemers committed:
Merge branch 'master' into anon-unknown-session-NEXUS-4257

bdemers committed:
NEXUS-4257,  Fixed problem with the happy anonymous login path.
Added happy path test. (A couple of the IT's found this problem, but only after a two hour build).

velo commented:
+1

peterlynch commented:
+1

bdemers commented:
Merged manually (to update the commit message)
